hvac/v1/__init__.py

```diff
@@ -265,6 +265,23 @@ def write(self, path, wrap_ttl=None, **kwargs):
         """
         return self._adapter.post(f"/v1/{path}", json=kwargs, wrap_ttl=wrap_ttl)
 
+    def write_fix_path(self, path, data={}, wrap_ttl=None):
```

Comment:
```suggestion
    def write_data(self, path, wrap_ttl=None, data={}):
```

Reply:
After reading it again maybe it's better to put `data` as mandatory like this `write_data(self, path, data, wrap_ttl=None)` because I don't see a case to write at a path without data. What do you think ?

Reply:
There are calls that need a write but all options are either optional or can be specified in the path, so I think having a default like that is valid. For example: https://developer.hashicorp.com/vault/api-docs/system/tools#generate-random-bytes

However, after reading it again, I actually agree with your original ordering 😅 and I would like to make `wrap_ttl` kw-only.
```suggestion
    def write_data(self, path, data={}, *, wrap_ttl=None):
```

I am also considering whether `data` should be kw-only as well 🤔 thoughts?

```suggestion
    def write_data(self, path, *, data={}, wrap_ttl=None):
```

I am thinking about how the calls might look with `data`:
```python
# no data, works either way
client.write_data("sys/tools/random")

# positional
data = {"format": "hex"}
client.write_data("sys/tools/random", {"format": "hex"})
client.write_data("sys/tools/random", dict(format="hex"))
client.write_data("sys/tools/random", data)

# keyword
client.write_data("sys/tools/random", data={"format": "hex"})
client.write_data("sys/tools/random", data=dict(format="hex")
client.write_data("sys/tools/random", data=data)
```

I'm leaning a little toward making everything except `path` be kw-only.

Reply:
I prefer the first suggested change, `wrap_ttl` can be an option that is not required and `data` is required but sometimes It can be an empty dict. Also what makes the most sense compared to the rest of the library ?

Reply:
When you say first suggested change, which one do you mean?

> Also what makes the most sense compared to the rest of the library?

One thing I have been seeing is that we get a lot of PRs to add new parameters to existing methods. For a lot of our methods, `mount_point` is the last parameter, and the impulse is to keep it last and add new parameters before it, or to rearrange parameters in alphabetical order when they correspond to the API parameters, etc.

The problem is that since we don't enforce keyword use, changing the order of parameters will break any existing uses that passed parameters positionally.

So as a result, I keep having to suggest to PR authors that they move the new parameters to the end of the method definition, even if that doesn't look right.

What I'd like to start doing in `hvac` is that we think of this intentionally when adding new methods, and going forward, we start a transition toward the (breaking) change of making most parameters of most methods kw-only, so that we can solve that going forward.

---

So that's where I'm coming from with this change. 

For this function, `path` makes sense to be the first parameter in this and many other calls, `data` _might_ make sense as a second positional (though I think that it might be better to force it as keyword), and anything else, I think should definitely be forced as a keyword.

Reply:
> That can happens only if hashicorp spec is updated. I mean when the signature of the function is created from the spec, if something change on their side it can breaks compatibility but yeah we need to takes it in consideration.

Sometimes this is because we missed parameters, or sometimes it's because new ones were added to the spec. But not all of the parameters in the python function directly correspond to the API spec, `mount_point` is a good example, since it is used as part of path generation.

> After reading again your example, data can be considered as an optional argument because this function is an abstraction of multiple API rest call that sometimes doesn't have DATA so we could consider it as a keyword argument.
`def write_data(self, path, *, data={}, wrap_ttl=None):` could be the best idea.

Ok, let's go with that then! Thank you.